bootstrap-joining-demo/aws-api/ec2-tag-based/src/it/scala/akka/cluster/bootstrap/IntegrationTest.scala

```diff
@@ -0,0 +1,233 @@
+import akka.http.scaladsl.model.HttpRequest
+import akka.management.cluster.{ClusterHttpManagementJsonProtocol, ClusterMembers}
+import akka.stream.{ActorMaterializer, ActorMaterializerSettings}
+import akka.util.ByteString
+import com.amazonaws.services.cloudformation.AmazonCloudFormationClientBuilder
+import com.amazonaws.services.cloudformation.model._
+import com.amazonaws.services.ec2.AmazonEC2ClientBuilder
+import com.amazonaws.services.ec2.model.{DescribeInstancesRequest, Filter, Instance, Reservation}
+import org.scalatest.concurrent.PatienceConfiguration.{Interval, Timeout}
+import org.scalatest.concurrent.{Eventually, ScalaFutures}
+import org.scalatest.time.{Seconds, Span, SpanSugar}
+import org.scalatest.{BeforeAndAfterAll, FunSuite}
+import spray.json._
+
+import scala.concurrent.{Await, Future}
+
+trait HttpClient {
+
+  implicit val system: ActorSystem = ActorSystem("simple")
+
+  implicit val materializer: ActorMaterializer = ActorMaterializer(ActorMaterializerSettings(system))
+
+  val http = Http(system)
+
+  import system.dispatcher
+
+  def httpGetRequest(url: String): Future[(Int, String)] =
+    http
+      .singleRequest(HttpRequest(uri = url))
+      .flatMap(
+          r =>
+            r.entity.dataBytes
+              .runFold(ByteString(""))(_ ++ _)
+              .map(_.utf8String)
+              .map(_.filter(_ >= ' '))
+              .map(body => (r.status.intValue(), body)))
+
+}
+
+class IntegrationTest
+    extends FunSuite
+    with Eventually
+    with BeforeAndAfterAll
+    with ScalaFutures
+    with HttpClient
+    with ClusterHttpManagementJsonProtocol
+    with SpanSugar
+{
+
+  import collection.JavaConverters._
+
+  private val buildId: String = System.getenv("BUILD_ID")
+  assert(buildId != null, "BUILD_ID environment variable has to be defined")
+
+  private val log = Logging(system, classOf[IntegrationTest])
+
+  private val instanceCount = 3
+
+  private val bucket = System.getenv("BUCKET") // bucket where zip file resulting from sbt universal:packageBin is stored
+
+  private val region = "us-east-1"
+
+  private val stackName = s"AkkaManagementIntegrationTestEC2TagBased-${buildId.replace(".", "-")}"
+
+  private val awsCfClient = AmazonCloudFormationClientBuilder.standard().withRegion(region).build()
+
+  private val awsEc2Client = AmazonEC2ClientBuilder.standard().withRegion(region).build()
+
+  // Patience settings for the part where we wait for the CloudFormation script to complete
+  private val createStackPatience: PatienceConfig =
+    PatienceConfig(
+      timeout = 15 minutes,
+      interval = 10 seconds
+    )
+
+  // Patience settings for the actual cluster bootstrap part.
+  // Once the CloudFormation stack has CREATE_COMPLETE status, the EC2 instances are
+  // still "initializing" (seems to take a very long time) so we add some additional patience for that.
+  private val clusterBootstrapPatience: PatienceConfig =
+    PatienceConfig(
+      timeout = 12 minutes,
```

Comment:
that is some patience :)

Reply:
@patriknw this is because it takes quite a few minutes for AWS to spin up instances (remember this is an integration test ...so actual EC2 instances are started ) ... I can reduce it from 12 minutes to something like 6 minutes, less than that it may be fail (not always , only from time to time )

Reply:
I don't mind it. better than false failures